File: src/dbgit/output_format.py
"""프로시저/함수 비교 결과를 텍스트·JSON·마크다운으로 직렬화 (CLI·UI 공통)."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Dict, Literal

if TYPE_CHECKING:
    from .compare import ProcDefinition

OutputFormat = Literal["text", "json", "markdown"]


def proc_comparison_payload(
    baseline: str,
    definitions: Dict[str, "ProcDefinition"],
) -> dict[str, object]:
    """JSON 직렬화 가능한 딕셔너리."""
    base_digest = definitions[baseline].digest
    env_entries: Dict[str, dict[str, object]] = {}
    for env_name, proc_def in definitions.items():
        env_entries[env_name] = {
            "object_id": proc_def.object_id,
            "schema_name": proc_def.schema_name,
            "name": proc_def.name,
            "full_name": proc_def.full_name,
            "digest": proc_def.digest,
            "same_as_baseline": proc_def.digest == base_digest,
        }
    diff_envs = [
        env_name
        for env_name, proc_def in definitions.items()
        if proc_def.digest != base_digest
    ]
    return {
        "baseline": baseline,
        "target_full_name": definitions[baseline].full_name,
        "environments": env_entries,
        "diff_environment_names": diff_envs,
        "all_same": len(diff_envs) == 0,
    }


def format_proc_comparison_text(baseline: str, definitions: Dict[str, "ProcDefinition"]) -> str:
    """기존 CLI 텍스트 형식."""
    base_def = definitions[baseline]
    lines: list[str] = []
    lines.append(f"기준 환경: {baseline} ({base_def.full_name})")
    lines.append("")
    lines.append("환경별 결과:")
    for env_name, proc_def in definitions.items():
        marker = "SAME" if proc_def.digest == base_def.digest else "DIFF"
        lines.append(f"- {env_name}: {marker} (object_id={proc_def.object_id})")
    lines.append("")
    diff_envs = [
        env_name for env_name, proc_def in definitions.items()
        if proc_def.digest != base_def.digest
    ]
    if diff_envs:
        lines.append("차이나는 환경:")
        lines.append(", ".join(diff_envs))
    else:
        lines.append("모든 환경이 동일합니다.")
    return "\n".join(lines)


def format_proc_comparison_json(baseline: str, definitions: Dict[str, "ProcDefinition"]) -> str:
    payload = proc_comparison_payload(baseline, definitions)
    definitions_dump: Dict[str, object] = {}
    for env_name, proc_def in definitions.items():
        definitions_dump[env_name] = {
            "object_id": proc_def.object_id,
            "schema_name": proc_def.schema_name,
            "name": proc_def.name,
            "full_name": proc_def.full_name,
            "digest": proc_def.digest,
            "definition": proc_def.definition,
        }
    payload["definitions_with_body"] = definitions_dump
    return json.dumps(payload, ensure_ascii=False, indent=2)


def format_proc_comparison_markdown(baseline: str, definitions: Dict[str, "ProcDefinition"]) -> str:
    base_def = definitions[baseline]
    lines: list[str] = [
        "## 프로시저·함수 비교 결과",
        "",
        f"- **기준 환경:** `{baseline}` — `{base_def.full_name}`",
        "",
        "| 환경 | 상태 | object_id | full_name |",
        "|:--|:--|--:|:--|",
    ]
    for env_name, proc_def in definitions.items():
        st = "SAME" if proc_def.digest == base_def.digest else "DIFF"
        lines.append(
            f"| {env_name} | {st} | {proc_def.object_id} | `{proc_def.full_name}` |"
        )
    diff_envs = [
        env_name for env_name, proc_def in definitions.items()
        if proc_def.digest != base_def.digest
    ]
    lines.append("")
    if diff_envs:
        lines.append(f"**차이 환경:** {', '.join(diff_envs)}")
    else:
        lines.append("**모든 환경이 동일합니다.**")
    return "\n".join(lines)


def format_proc_comparison(
    baseline: str,
    definitions: Dict[str, "ProcDefinition"],
    fmt: OutputFormat,
) -> str:
    if fmt == "json":
        return format_proc_comparison_json(baseline, definitions)
    if fmt == "markdown":
        return format_proc_comparison_markdown(baseline, definitions)
    return format_proc_comparison_text(baseline, definitions)

File: src/outputFormat.ts
/**
 * 출력 포맷(text/json/markdown).
 *
 * 비교 결과를 사람이 보기 좋게 렌더링합니다.
 */
import type { ProcDefinition } from "./compare.js";
import { digestOf } from "./compare.js";

export type OutputFormat = "text" | "json" | "markdown";

function diffEnvNames(baseline: string, definitions: Record<string, ProcDefinition>): string[] {
  const baseDigest = digestOf(definitions[baseline]!);
  return Object.entries(definitions)
    .filter(([, d]) => digestOf(d) !== baseDigest)
    .map(([n]) => n);
}

export function formatProcComparisonText(
  baseline: string,
  definitions: Record<string, ProcDefinition>
): string {
  const baseDef = definitions[baseline]!;
  const lines: string[] = [];
  lines.push(`기준 환경: ${baseline} (${baseDef.schema_name}.${baseDef.name})`);
  lines.push("");
  lines.push("환경별 결과:");
  for (const [envName, procDef] of Object.entries(definitions)) {
    const marker = digestOf(procDef) === digestOf(baseDef) ? "SAME" : "DIFF";
    lines.push(`- ${envName}: ${marker} (object_id=${procDef.object_id})`);
  }
  lines.push("");
  const diffEnvs = diffEnvNames(baseline, definitions);
  if (diffEnvs.length) {
    lines.push("차이나는 환경:");
    lines.push(diffEnvs.join(", "));
  } else {
    lines.push("모든 환경이 동일합니다.");
  }
  return lines.join("\n");
}

export function formatProcComparisonJson(
  baseline: string,
  definitions: Record<string, ProcDefinition>
): string {
  const baseDigest = digestOf(definitions[baseline]!);
  const envEntries: Record<string, object> = {};
  for (const [envName, procDef] of Object.entries(definitions)) {
    envEntries[envName] = {
      object_id: procDef.object_id,
      schema_name: procDef.schema_name,
      name: procDef.name,
      full_name: `${procDef.schema_name}.${procDef.name}`,
      digest: digestOf(procDef),
      same_as_baseline: digestOf(procDef) === baseDigest,
    };
  }
  const diff_environment_names = diffEnvNames(baseline, definitions);
  const payload = {
    baseline,
    target_full_name: `${definitions[baseline]!.schema_name}.${definitions[baseline]!.name}`,
    environments: envEntries,
    diff_environment_names,
    all_same: diff_environment_names.length === 0,
    definitions_with_body: Object.fromEntries(
      Object.entries(definitions).map(([k, v]) => [
        k,
        {
          object_id: v.object_id,
          schema_name: v.schema_name,
          name: v.name,
          full_name: `${v.schema_name}.${v.name}`,
          digest: digestOf(v),
          definition: v.definition,
        },
      ])
    ),
  };
  return JSON.stringify(payload, null, 2);
}

export function formatProcComparisonMarkdown(
  baseline: string,
  definitions: Record<string, ProcDefinition>
): string {
  const baseDef = definitions[baseline]!;
  const fn = `${baseDef.schema_name}.${baseDef.name}`;
  const lines: string[] = [
    "## 프로시저·함수 비교 결과",
    "",
    `- **기준 환경:** \`${baseline}\` — \`${fn}\``,
    "",
    "| 환경 | 상태 | object_id | full_name |",
    "|:--|:--|--:|:--|",
  ];
  for (const [envName, procDef] of Object.entries(definitions)) {
    const st = digestOf(procDef) === digestOf(baseDef) ? "SAME" : "DIFF";
    lines.push(`| ${envName} | ${st} | ${procDef.object_id} | \`${procDef.schema_name}.${procDef.name}\` |`);
  }
  lines.push("");
  const diffEnvs = diffEnvNames(baseline, definitions);
  if (diffEnvs.length) {
    lines.push(`**차이 환경:** ${diffEnvs.join(", ")}`);
  } else {
    lines.push("**모든 환경이 동일합니다.**");
  }
  return lines.join("\n");
}

export function formatProcComparison(
  baseline: string,
  definitions: Record<string, ProcDefinition>,
  fmt: OutputFormat
): string {
  if (fmt === "json") return formatProcComparisonJson(baseline, definitions);
  if (fmt === "markdown") return formatProcComparisonMarkdown(baseline, definitions);
  return formatProcComparisonText(baseline, definitions);
}
